File: ts/src/ik/kinematics.ts
// SPDX-FileCopyrightText: 2026 Mario Gemoll
// SPDX-License-Identifier: 0BSD

import * as THREE from 'three';

import {
  GRIPPER_TARGET_POSITION
} from '../visualizations/grasp-pose-shared/body-factories';
import type { WebJoint, WebModel } from '../web-model';

// The five actuated joints of the SO-101 arm, base to tool, in chain order.
export const ARM_JOINT_NAMES = [
  'shoulder_pan',
  'shoulder_lift',
  'elbow_flex',
  'wrist_flex',
  'wrist_roll'
] as const;

export type ArmJointName = typeof ARM_JOINT_NAMES[number];

export const NEUTRAL_ARM_JOINTS: Record<ArmJointName, number> = {
  shoulder_pan: 0,
  shoulder_lift: 0,
  elbow_flex: 0,
  wrist_flex: 0,
  wrist_roll: -Math.PI / 2
};

// A link of the planar arm, expressed in the radial–height plane that the
// closed-form IK solves in. `radial` runs outward from the pan axis, `height`
// is world z; the link's constant out-of-plane (lateral) offset is dropped, so
// `length` is the in-plane projection that the 2R solver treats as a link.
export interface PlanarSegment {
  radial: number;
  height: number;
  length: number;
}

export interface JointLimit {
  min: number;
  max: number;
}

// Kinematic constants the closed-form IK needs, all derived from the loaded
// model so they stay in sync with `so101.json`.
// Lengths are metres, limits radians.
export interface So101Kinematics {
  // Horizontal location of the vertical shoulder_pan axis (world x, y).
  panAxis: THREE.Vector2;
  // shoulder_lift pivot in the radial–height plane: radial offset from the pan
  // axis, height above the floor.
  shoulderLift: { radial: number; height: number };
  upperArm: PlanarSegment; // shoulder_lift -> elbow_flex
  lowerArm: PlanarSegment; // elbow_flex -> wrist_flex
  // wrist_flex -> gripper target along the approach direction; the tool that
  // the wrist pitch carries.
  toolLength: number;
  // The constant roll angle offset (in radians) of the gripper x-axis relative
  // to the ideal zero-roll pitch axis (due to the 2.8° arm twist).
  wristRollZeroTwist: number;
  // Joint limits for the five arm joints.
  jointLimits: Record<ArmJointName, JointLimit>;
}

function quaternionFromWeb(
  [w, x, y, z]: [number, number, number, number]
): THREE.Quaternion {
  return new THREE.Quaternion(x, y, z, w);
}

// World transform of every body at the zero pose. Bodies are listed parent
// before child, so a single forward pass suffices.
function worldMatrices(model: WebModel): Map<string, THREE.Matrix4> {
  const worlds = new Map<string, THREE.Matrix4>();
  for (const body of model.bodies) {
    const local = new THREE.Matrix4().compose(
      new THREE.Vector3(...body.position),
      quaternionFromWeb(body.quaternion),
      new THREE.Vector3(1, 1, 1)
    );
    const parentWorld = worlds.get(body.parent);
    worlds.set(
      body.name,
      parentWorld ? parentWorld.clone().multiply(local) : local
    );
  }
  return worlds;
}

interface JointFrame {
  position: THREE.Vector3;
  axis: THREE.Vector3;
  joint: WebJoint;
}

function jointFrame(
  model: WebModel,
  worlds: Map<string, THREE.Matrix4>,
  name: ArmJointName
): JointFrame {
  for (const body of model.bodies) {
    const joint = body.joints.find(candidate => candidate.name === name);
    if (joint === undefined) { continue; }
    const world = worlds.get(body.name);
    if (world === undefined) { break; }
    return {
      position: new THREE.Vector3(...joint.position).applyMatrix4(world),
      axis: new THREE.Vector3(...joint.axis).transformDirection(world).normalize(),
      joint
    };
  }
  throw new Error(`Joint ${name} not found in model`);
}

function jointLimit(joint: WebJoint): JointLimit {
  if (joint.limited && joint.range) {
    return { min: joint.range[0], max: joint.range[1] };
  }
  return { min: -Infinity, max: Infinity };
}

export function deriveSo101Kinematics(model: WebModel): So101Kinematics {
  const worlds = worldMatrices(model);
  const pan = jointFrame(model, worlds, 'shoulder_pan');
  const lift = jointFrame(model, worlds, 'shoulder_lift');
  const elbow = jointFrame(model, worlds, 'elbow_flex');
  const wristFlex = jointFrame(model, worlds, 'wrist_flex');
  const wristRoll = jointFrame(model, worlds, 'wrist_roll');

  const panAxis = new THREE.Vector2(pan.position.x, pan.position.y);

  // The radial axis is horizontal and perpendicular to the (lateral) pitch
  // axis: pitch × up, projected to the floor and oriented outward toward the
  // arm. This drops the arm's constant lateral offset automatically.
  const radialDir = new THREE.Vector3()
    .crossVectors(lift.axis, new THREE.Vector3(0, 0, 1));
  const radial = new THREE.Vector2(radialDir.x, radialDir.y).normalize();
  const toWrist = new THREE.Vector2(
    wristFlex.position.x - panAxis.x,
    wristFlex.position.y - panAxis.y
  );
  if (radial.dot(toWrist) < 0) { radial.negate(); }

  const radialOf = (position: THREE.Vector3): number =>
    (position.x - panAxis.x) * radial.x + (position.y - panAxis.y) * radial.y;
  const segment = (
    from: THREE.Vector3,
    to: THREE.Vector3
  ): PlanarSegment => {
    const dr = radialOf(to) - radialOf(from);
    const dh = to.z - from.z;
    return { radial: dr, height: dh, length: Math.hypot(dr, dh) };
  };

  const target = GRIPPER_TARGET_POSITION.clone()
    .applyMatrix4(worlds.get('gripper') ?? new THREE.Matrix4());

  const gripperWorld = worlds.get('gripper') ?? new THREE.Matrix4();
  const gripperX = new THREE.Vector3(1, 0, 0).transformDirection(gripperWorld).normalize();
  const a = new THREE.Vector3().subVectors(target, wristFlex.position).normalize();
  // Pitch axis n_V = (0, 1, 0) at pan=0 (normal to the radial plane).
  const pitchAxis = new THREE.Vector3(0, 1, 0);
  const idealX = new THREE.Vector3().crossVectors(a, pitchAxis).normalize();
  const idealY = pitchAxis;
  const wristRollZeroTwist = Math.atan2(gripperX.dot(idealY), gripperX.dot(idealX));

  return {
    panAxis,
    shoulderLift: { radial: radialOf(lift.position), height: lift.position.z },
    upperArm: segment(lift.position, elbow.position),
    lowerArm: segment(elbow.position, wristFlex.position),
    toolLength: segment(wristFlex.position, target).length,
    wristRollZeroTwist,
    jointLimits: {
      shoulder_pan: jointLimit(pan.joint),
      shoulder_lift: jointLimit(lift.joint),
      elbow_flex: jointLimit(elbow.joint),
      wrist_flex: jointLimit(wristFlex.joint),
      wrist_roll: jointLimit(wristRoll.joint)
    }
  };
}

File: py/src/pick_and_place/kinematics.py
# SPDX-FileCopyrightText: 2026 Mario Gemoll
# SPDX-License-Identifier: 0BSD

"""SO-101 kinematic constants for the closed-form IK.

The constants are read straight off the compiled MuJoCo model at its reference
(all-joints-zero) pose, which exposes joint world anchors and axes directly via
``mjData``.
"""

from __future__ import annotations

from dataclasses import dataclass

import mujoco
import numpy as np

from pick_and_place.geometry import GRIPPER_TARGET_POSITION
from pick_and_place.transforms import Vec3

ARM_JOINT_NAMES = (
    "shoulder_pan",
    "shoulder_lift",
    "elbow_flex",
    "wrist_flex",
    "wrist_roll",
)


@dataclass(frozen=True)
class PlanarSegment:
    radial: float
    height: float
    length: float


@dataclass(frozen=True)
class JointLimit:
    min: float
    max: float


@dataclass(frozen=True)
class So101Kinematics:
    pan_axis: Vec3  # world (x, y) of the vertical shoulder_pan axis; z unused
    shoulder_lift_radial: float
    shoulder_lift_height: float
    upper_arm: PlanarSegment  # shoulder_lift -> elbow_flex
    lower_arm: PlanarSegment  # elbow_flex -> wrist_flex
    tool_length: float  # wrist_flex -> gripper target along the approach
    wrist_roll_zero_twist: float  # roll offset from the 2.8 deg arm twist
    joint_limits: dict[str, JointLimit]

    def tip_position(self, joints: dict[str, float]) -> Vec3:
        """World position of the gripper IK target for the given arm joints.

        Closed-form forward kinematics of the planar chain, inverting the
        conventions in ``solve_simple_grasp_ik``. Wrist roll spins about the
        approach axis and so does not move the target; only pan and the three
        planar joints matter. Matches the IK target to sub-millimetre.
        """
        azimuth = -joints["shoulder_pan"]
        radial_dir = np.array((np.cos(azimuth), np.sin(azimuth), 0.0))
        up = np.array((0.0, 0.0, 1.0))
        upper_rest = np.arctan2(self.upper_arm.height, self.upper_arm.radial)
        lower_rest = np.arctan2(self.lower_arm.height, self.lower_arm.radial)
        elbow_rest = lower_rest - upper_rest
        shoulder_geom = upper_rest - joints["shoulder_lift"]
        elbow_geom = elbow_rest - joints["elbow_flex"]
        l1, l2 = self.upper_arm.length, self.lower_arm.length
        radial_rel = l1 * np.cos(shoulder_geom) + l2 * np.cos(shoulder_geom + elbow_geom)
        height_rel = l1 * np.sin(shoulder_geom) + l2 * np.sin(shoulder_geom + elbow_geom)
        wrist = (
            np.array((self.pan_axis[0], self.pan_axis[1], 0.0))
            + radial_dir * (self.shoulder_lift_radial + radial_rel)
            + up * (self.shoulder_lift_height + height_rel)
        )
        tool_pitch = -(joints["shoulder_lift"] + joints["elbow_flex"] + joints["wrist_flex"])
        approach = radial_dir * np.cos(tool_pitch) + up * np.sin(tool_pitch)
        return wrist + approach * self.tool_length


@dataclass(frozen=True)
class _JointFrame:
    position: Vec3
    axis: Vec3


def _joint_frame(model: mujoco.MjModel, data: mujoco.MjData, name: str) -> _JointFrame:
    jid = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_JOINT, name)
    return _JointFrame(position=data.xanchor[jid].copy(), axis=data.xaxis[jid].copy())


def _joint_limit(model: mujoco.MjModel, name: str) -> JointLimit:
    jid = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_JOINT, name)
    if model.jnt_limited[jid]:
        lo, hi = model.jnt_range[jid]
        return JointLimit(float(lo), float(hi))
    return JointLimit(-np.inf, np.inf)


def derive_kinematics(model: mujoco.MjModel) -> So101Kinematics:
    data = mujoco.MjData(model)
    data.qpos[:] = model.qpos0
    mujoco.mj_forward(model, data)

    pan = _joint_frame(model, data, "shoulder_pan")
    lift = _joint_frame(model, data, "shoulder_lift")
    elbow = _joint_frame(model, data, "elbow_flex")
    wrist_flex = _joint_frame(model, data, "wrist_flex")

    pan_axis = pan.position.copy()

    # Radial axis: horizontal, perpendicular to the (lateral) pitch axis, oriented
    # outward toward the arm. lift.axis x up, projected to the floor.
    radial_dir = np.cross(lift.axis, np.array((0.0, 0.0, 1.0)))
    radial = np.array((radial_dir[0], radial_dir[1]))
    radial /= np.linalg.norm(radial)
    to_wrist = wrist_flex.position[:2] - pan_axis[:2]
    if float(np.dot(radial, to_wrist)) < 0:
        radial = -radial

    def radial_of(position: Vec3) -> float:
        d = position[:2] - pan_axis[:2]
        return float(d[0] * radial[0] + d[1] * radial[1])

    def segment(frm: Vec3, to: Vec3) -> PlanarSegment:
        dr = radial_of(to) - radial_of(frm)
        dh = float(to[2] - frm[2])
        return PlanarSegment(radial=dr, height=dh, length=float(np.hypot(dr, dh)))

    gid = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, "gripper")
    gripper_pos = data.xpos[gid].copy()
    gripper_rot = data.xmat[gid].reshape(3, 3)
    target = gripper_pos + gripper_rot @ GRIPPER_TARGET_POSITION

    gripper_x = gripper_rot[:, 0] / np.linalg.norm(gripper_rot[:, 0])
    a = target - wrist_flex.position
    a /= np.linalg.norm(a)
    pitch_axis = np.array((0.0, 1.0, 0.0))  # holds at pan = 0 (reference pose)
    ideal_x = np.cross(a, pitch_axis)
    ideal_x /= np.linalg.norm(ideal_x)
    ideal_y = pitch_axis
    wrist_roll_zero_twist = float(
        np.arctan2(np.dot(gripper_x, ideal_y), np.dot(gripper_x, ideal_x))
    )

    return So101Kinematics(
        pan_axis=pan_axis,
        shoulder_lift_radial=radial_of(lift.position),
        shoulder_lift_height=float(lift.position[2]),
        upper_arm=segment(lift.position, elbow.position),
        lower_arm=segment(elbow.position, wrist_flex.position),
        tool_length=segment(wrist_flex.position, target).length,
        wrist_roll_zero_twist=wrist_roll_zero_twist,
        joint_limits={name: _joint_limit(model, name) for name in ARM_JOINT_NAMES},
    )
